jlm/util/BiMap.hpp

```diff
@@ -0,0 +1,254 @@
+   * @param value the value to look for
+   * @returns true if the value exists in the BiMap, otherwise false.
+   */
+  [[nodiscard]] bool
+  HasValue(const V & value) const noexcept
+  {
+    return ReverseMap_.find(value) != ReverseMap_.end();
+  }
+
+  /**
+   * @param key the key to lookup in the BiMap
+   * @returns the value associated with the given \p key
+   * @throws jlm::util::error if the \p key does not exist in the BiMap
+   */
+  [[nodiscard]] const V &
+  LookupKey(const K & key) const
+  {
+    auto it = ForwardMap_.find(key);
+    if (it == ForwardMap_.end())
+      throw jlm::util::error("Key not found in BiMap");
+    return it->second;
+  }
+
+  /**
+   * @param value the value to lookup in the BiMap
+   * @returns the key associated with the given \p value
+   * @throws jlm::util::error if the \p value does not exist in the BiMap
+   */
+  [[nodiscard]] const K &
+  LookupValue(const V & value) const
+  {
+    auto it = ReverseMap_.find(value);
+    if (it == ReverseMap_.end())
+      throw jlm::util::error("Value not found in BiMap");
+    return it->second;
+  }
+
+  /**
+   * Removes a key/value pair from the BiMap, if the given \p key exists
+   * @param key the key to be removed
+   * @returns true if a key/value pair was removed
+   */
+  bool
+  RemoveKey(const K & key)
+  {
+    auto it = ForwardMap_.find(key);
+    if (it == ForwardMap_.end())
+      return false;
+
+    auto removed = ReverseMap_.erase(it->second);
+    JLM_ASSERT(removed == 1);
+
+    ForwardMap_.erase(it);
+    return true;
+  }
+
+  /**
+   * Removes a key/value pair from the BiMap, if the given \p value exists
+   * @param value the value to be removed
+   * @returns true if a key/value pair was removed
+   */
+  bool
+  RemoveValue(const V & value)
+  {
+    auto it = ReverseMap_.find(value);
+    if (it == ReverseMap_.end())
+      return false;
+
+    auto removed = ForwardMap_.erase(it->second);
+    JLM_ASSERT(removed == 1);
+
+    ReverseMap_.erase(it);
+    return true;
+  }
+
+  /**
+   * @returns the underlying mapping from keys to values
+   */
+  [[nodiscard]] const std::unordered_map<K, V> &
+  GetForwardMap() const noexcept
+  {
+    return ForwardMap_;
+  }
+
+  /**
+   * @returns the underlying mapping from values to keys
+   */
+  [[nodiscard]] const std::unordered_map<V, K> &
+  GetReverseMap() const noexcept
+  {
+    return ReverseMap_;
+  }
+
+  /**
+   * Shorthand for iterating over all key/value pairs in the BiMap
+   * @returns a const iterator pointing to the first key/value pair in the ForwardMap
+   */
+  [[nodiscard]] typename std::unordered_map<K, V>::const_iterator
+  begin() const noexcept
+  {
+    return ForwardMap_.cbegin();
+  }
+
+  /**
+   * Shorthand for iterating over all key/value pairs
+   * @returns a const iterator pointing to the end of the ForwardMap
+   */
+  [[nodiscard]] typename std::unordered_map<K, V>::const_iterator
+  end() const noexcept
+  {
+    return ForwardMap_.cend();
+  }
```

Comment:
The same here. The returned `const_iterator`s expose the implementation. Could we wrap them in your own BijectiveMap::ConstIterator?